Per-partition rate limiting

piodul opened the pull request:
Due to its sharded and token-based architecture, Scylla works best when the user workload is more or less uniformly balanced across all nodes and shards. However, a common case when this assumption is broken is the "hot partition" - suddenly, a single partition starts getting a lot more reads and writes in comparison to other partitions. Because the shards owning the partition have only a fraction of the total cluster capacity, this quickly causes latency problems for other partitions within the same shard and vnode.

This PR introduces per-partition rate limiting feature. Now, users can choose to apply per-partition limits to their tables of choice using a schema extension:

```
ALTER TABLE ks.tbl
WITH per_partition_rate_limit = {
	'max_writes_per_second': 100,
	'max_reads_per_second': 200
};
```

Reads and writes which are detected to go over that quota are rejected to the client using a new RATE_LIMIT_ERROR CQL error code - existing error codes didn't really fit well with the rate limit error, so a new error code is added. This code is implemented as a part of a CQL protocol extension and returned to clients only if they requested the extension - if not, the existing CONFIG_ERROR will be used instead.

Limits are tracked and enforced on the replica side. If a write fails with some replicas reporting rate limit being reached, the rate limit error is propagated to the client. Additionally, the following optimization is implemented: if the coordinator shard/node is also a replica, we account the operation into the rate limit early and return an error in case of exceeding the rate limit before sending any messages to other replicas at all.

The PR covers regular, non-batch writes and single-partition reads. LWT and counters are not covered here.

Results of `perf_simple_query --smp=1 --operations-per-shard=1000000`:

- Write mode:
  ```
  8f690fdd47ba840360ca38620359be9a3ae09c58 (PR base):
  129644.11 tps ( 56.2 allocs/op,  13.2 tasks/op,   49785 insns/op)
  This PR:
  125564.01 tps ( 56.2 allocs/op,  13.2 tasks/op,   49825 insns/op)
  ```
- Read mode:
  ```
  8f690fdd47ba840360ca38620359be9a3ae09c58 (PR base):
  150026.63 tps ( 63.1 allocs/op,  12.1 tasks/op,   42806 insns/op)
  This PR:
  151043.00 tps ( 63.1 allocs/op,  12.1 tasks/op,   43075 insns/op)
  ```

Manual upgrade test:
- Start 3 nodes, 4 shards each, Scylla version 8f690fdd47ba840360ca38620359be9a3ae09c58
- Create a keyspace with scylla-bench, RF=3
- Start reading and writing with scylla-bench with CL=QUORUM
- Manually upgrade nodes one by one to the version from this PR
- Upgrade succeeded, apart from a small number of operations which failed when each node was being put down all reads/writes succeeded
- Successfully altered the scylla-bench table to have a read and write limit and those limits were enforced as expected

Fixes: #4703

piodul commented:
I wanted to perform more extensive benchmarks, but I got some weird behavior from my cluster and didn't have time to investigate.

In the small benchmark below, I used a two-node cluster set up on two Warsaw Office's machines, with 1 shard and 32G memory each. CL=ONE and RF=2 was used.

Here are some graphs which show the rate limiting in action. I ran a scylla-bench compatible tool which wrote very small mutations to a single partition, 10k writes/s. Halfway through the test, I limited the allowed per-partition writes to 100/s. The resulting CPU usage dropped very slightly:

![Zrzut ekranu z 2021-12-16 01-30-29](https://user-images.githubusercontent.com/4950936/146360551-c663074c-6a1c-49f2-940a-831529fa6fc2.png)

The small bumps which happen every 10 seconds are caused by hinted handoff replay. This is due to the time window based enforcement of the rate limits - both replicas has unsynchronized time windows, so there were small periods of time when one replica rejected writes but the other one still accepted them.

Although the `Load` metric seems to go down when rate limiting is enabled, when increasing the number of requests per second the results aren't so good anymore and latency increases when rate limiting is enabled. I suspect that it is due to the exception throwing/rethrowing on the coordinator (the `_Unwind_XYZ` functions appeared in the flamegraphs), but I will have to investigate in more detail after coming back from vacation.

gleb-cloudius commented on message/messaging_service.cc:
```diff
@@ -1046,14 +1046,14 @@ future<> messaging_service::send_mutation_done(msg_addr id, unsigned shard, resp
 }
 
-void messaging_service::register_mutation_failed(std::function<future<rpc::no_wait_type> (const rpc::client_info& cinfo, unsigned shard, response_id_type response_id, size_t num_failed, rpc::optional<db::view::update_backlog> backlog)>&& func) {
+void messaging_service::register_mutation_failed(std::function<future<rpc::no_wait_type> (const rpc::client_info& cinfo, unsigned shard, response_id_type response_id, size_t num_failed, rpc::optional<db::view::update_backlog> backlog, bool rate_limit_hit)>&& func) {
```
All extensions should be added as rpc::optional<>. Also make it even more generic then enum. Something like:
```
struct error {
  struct rate_limited{}; 
  std::variant<rate_limited> reason;
}
```

psarna commented:
@piodul plz post a changelog for v2 to facilitate the review process

denesb reviewed (commented):
I have a few comments?
* I see we have an explicit flag to turn the rate-limiting feature on on the replica. Under what circumstances is this ever disabled (I only saw ::yes used).
* I saw that this is controlled in the form of a schema extension. I just want to make sure we don't turn it on for system reads. What if a user enables it on a system table? How do we make sure system (internal) reads are still not affected?

piodul commented:
> @piodul plz post a changelog for v2 to facilitate the review process

(There are some things still missing and I'll push them soon after I implement them)

v0.1 -> v0.2 changes:
- Rebased onto newer master branch
- Better documentation in db::rate_limiter and implementation improvements
- Rate limiting reads is implemented
- More comprehensive approach to passing the allow_per_partition_rate_limit flag
- Errors are propagated through RPC based on https://github.com/scylladb/scylla/pull/10330 (TODO: do not convert existing errors but just the rate_limit_exception)

TODO:
- Optimization: abort read/write early if the coordinator detects that the limit is reached
- Approach the exception propagation problem (resultization || catch without throwing)
- Handle the problem of rate_limit accepting reads/writes in bursts

piodul commented on replica/database.cc:
```diff
@@ -1342,17 +1354,28 @@ database::query(schema_ptr s, const query::read_command& cmd, query::result_opti
 
+    column_family& cf = find_column_family(cmd.cf_id);
+
+    if (allow_limit && range.is_singular()) {
```
Right, I admit that I added the flag here without thinking it through. I'll remove it in the next version.

piodul replied:
Done, now read repair shouldn't be rate limited.

piodul commented on replica/database.cc:
```diff
@@ -1768,6 +1791,16 @@ future<> database::do_apply(schema_ptr s, const frozen_mutation& m, tracing::tra
     }
 
+    if (bool(allow_limit)) {
+      if (auto table_limit = s->per_partition_rate_limit_options().get_max_writes_per_second()) {
```
This change used to be introduced in two commits but I squashed them during cleanup and I forgot to un-half-step it. I'll correct it in the next version.

piodul replied:
Fixed

piodul commented:
> I have a few comments:
> 
> * I see we have an explicit flag to turn the rate-limiting feature on on the replica. Under what circumstances is this ever disabled (I only saw ::yes used).

@denesb I'm already using `::no` already in some places, for example:
- CDC generation-related functions in `system_distributed_keyspace.cc` operate on system tables and explicitly disable the rate limit,
- CDC preimage query in `cdc/log.cc` is not explicitly issued by the user so it's not covered by the rate limit.

> * I saw that this is controlled in the form of a schema extension. I just want to make sure we don't turn it on for system reads. What if a user enables it on a system table? How do we make sure system (internal) reads are still not affected?

I must admit that I don't have a nice idea how to solve this (I'm open to suggestions if anybody has a better idea).

I approached it by adding the allow_per_partition_rate_limit parameter to the relevant methods in storage_proxy and made it mandatory (I think I actually made it optional in some of the methods but I'll make it required). The responsibility of deciding whether it's a good idea to enforce the rate limit is pushed to the caller and they need to be explicit about it. Internal operations on system tables and some operations which are not directly issued by the user (e.g. read repair, hints) should explicitly disable it.

The user still probably can ALTER a system table to use rate limiting just because they can, but internal reads shouldn't be affected.

denesb commented:
> > I have a few comments:
> > 
> > * I see we have an explicit flag to turn the rate-limiting feature on on the replica. Under what circumstances is this ever disabled (I only saw ::yes used).
> 
> @denesb I'm already using `::no` already in some places, for example:
> 
>     * CDC generation-related functions in `system_distributed_keyspace.cc` operate on system tables and explicitly disable the rate limit,
> 
>     * CDC preimage query in `cdc/log.cc` is not explicitly issued by the user so it's not covered by the rate limit.
> 
> 
> > * I saw that this is controlled in the form of a schema extension. I just want to make sure we don't turn it on for system reads. What if a user enables it on a system table? How do we make sure system (internal) reads are still not affected?
> 
> I must admit that I don't have a nice idea how to solve this (I'm open to suggestions if anybody has a better idea).
> 
> I approached it by adding the allow_per_partition_rate_limit parameter to the relevant methods in storage_proxy and made it mandatory (I think I actually made it optional in some of the methods but I'll make it required). The responsibility of deciding whether it's a good idea to enforce the rate limit is pushed to the caller and they need to be explicit about it. Internal operations on system tables and some operations which are not directly issued by the user (e.g. read repair, hints) should explicitly disable it.
> 
> The user still probably can ALTER a system table to use rate limiting just because they can, but internal reads shouldn't be affected.

Most internal queries are not executed with calling storage proxy methods directly, rather they are regular CQL queries executed via query processor. One way to distinguish between user and system queries is to look at the scheduling group they run in. This is the approach used by selecting the appropriate query size limit and semaphore, see `classify_query()` in `database.cc`. Maybe you can use this instead of propagating a flag. Note however that system queries run on behalf of a user query (and that therefore run in the `statement` scheduling group) will be classified by user queries by this logic.

denesb commented:
> * I just want to make sure we don't turn it on for system reads. What if a user enables it on a system table?

To be clear it *is fine* for the user to enable it on a system table, so long as it only affects user reads. Although this can still cause problems. Drivers will read system tables and an ill-setup limit on the `local` key of `system.local` can completely kill all client connections.

piodul commented on replica/database.cc:
```diff
@@ -1283,13 +1284,24 @@ database::existing_index_names(const sstring& ks_name, const sstring& cf_to_excl
+    if (allow_limit && ranges.front().is_singular()) {
+        if (auto table_limit = s->per_partition_rate_limit_options().get_max_reads_per_second()) {
+            auto& read_label = cf.get_rate_limiter_label_for_reads();
+            auto token = dht::token::to_int64(ranges.front().start()->value().token());
```
I changed the logic so that all singular ranges are accounted.

However, I think there is another problem with my current implementation. Now, replicas reject reads with some probability and for each singular range the decision is made separately. A read will be rejected if the decision is "reject" for _any_ of the partitions, so the combined probability of rejection is higher than any separate partition.

I'm not sure how should we behave in this case. Compute the probabilities of rejection, take maximum and then toss a coin based on that? Or maybe not rate limit at all, to be consistent with scans (for which I don't think there is a sensible way to apply per-partition rate limit)?

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/469/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/471/

avikivity commented on transport/server.cc:
```diff
@@ -1275,6 +1278,15 @@ std::unique_ptr<cql_server::response> cql_server::connection::make_function_fail
+    if (!client_state.is_protocol_extension_set(cql_protocol_extension::RATE_LIMIT_ERROR)) {
+        return make_error(stream, exceptions::exception_code::CONFIG_ERROR, std::move(msg), tr_state);
+    }
+
```
Very good! This both takes care of compatibility and gives fine-grained information for updated drivers.

avikivity commented on idl/storage_proxy.idl.hh:
```diff
@@ -8,7 +8,7 @@
 verb [[with_client_info, with_timeout, one_way]] mutation (frozen_mutation fm, inet_address_vector_replica_set forward, gms::inet_address reply_to, unsigned shard, uint64_t response_id, std::optional<tracing::trace_info> trace_info [[version 1.3.0]]);
 verb [[with_client_info, one_way]] mutation_done (unsigned shard, uint64_t response_id, db::view::update_backlog backlog [[version 3.1.0]]);
-verb [[with_client_info, one_way]] mutation_failed (unsigned shard, uint64_t response_id, size_t num_failed, db::view::update_backlog backlog [[version 3.1.0]]);
+verb [[with_client_info, one_way]] mutation_failed (unsigned shard, uint64_t response_id, size_t num_failed, db::view::update_backlog backlog [[version 3.1.0]], replica::exception_variant exception [[version 5.1.0]]);
```
Please verify that mixed clusters indeed work, we usually use `rpc::optional<> ` for extensions. If it works without it, that's preferable.

gleb-cloudius replied:
This is idl. `rpc::optional` will be added by the idl compiler.

avikivity commented on idl/storage_proxy.idl.hh:
```diff
@@ -11,9 +11,9 @@ verb [[with_client_info, one_way]] mutation_done (unsigned shard, uint64_t respo
-verb [[with_client_info, with_timeout]] read_mutation_data (query::read_command cmd, ::compat::wrapping_partition_range pr) -> reconcilable_result [[lw_shared_ptr]], cache_temperature [[version 2.0.0]];
-verb [[with_client_info, with_timeout]] read_digest (query::read_command cmd, ::compat::wrapping_partition_range pr, query::digest_algorithm digest [[version 3.0.0]]) -> query::result_digest, api::timestamp_type [[version 1.2.0]], cache_temperature [[version 2.0.0]];
+verb [[with_client_info, with_timeout]] read_data (query::read_command cmd, ::compat::wrapping_partition_range pr, query::digest_algorithm digest [[version 3.0.0]]) -> query::result [[lw_shared_ptr]], cache_temperature [[version 2.0.0]], std::optional<replica::exception_variant> [[version 5.1.0]];
+verb [[with_client_info, with_timeout]] read_mutation_data (query::read_command cmd, ::compat::wrapping_partition_range pr) -> reconcilable_result [[lw_shared_ptr]], cache_temperature [[version 2.0.0]], std::optional<replica::exception_variant> [[version 5.1.0]];
```
IIRC the marker is `rpc::optional`.

gleb-cloudius replied:
Why `std::optional`? exception_variant can be empty (monostate) which may indicate "no error" instead of optional.

gleb-cloudius replied:
You can make exception class to contain no exception by adding one more state `no_exception`.  Sort of make it optional by itself.

piodul replied:
Done, added the `no_exception` variant.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -2989,18 +2997,22 @@ class digest_read_resolver : public abstract_read_resolver {
         }
         if (_block_for + _failed > _target_count_for_cl) {
-            fail_request(read_failure_exception(_schema->ks_name(), _schema->cf_name(), _cl, _cl_responses, _failed, _block_for, _data_result));
+            if (kind == error_kind::RATE_LIMIT) {
```
It's better to use `switch` (and no `default`) for enums so adding an enum member is safer as the compiler forces us to check call the sites.

piodul replied:
Done

avikivity commented on service/storage_proxy.cc:
```diff
@@ -3581,8 +3600,11 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
-                auto&& [result, hit_rate] = result_hit_rate;
+            return ser::storage_proxy_rpc_verbs::send_read_data(&_proxy->_messaging, netw::messaging_service::msg_addr{ep, 0}, timeout, *_cmd, _partition_range, opts.digest_algo).then([this, ep](rpc::tuple<query::result, rpc::optional<cache_temperature>, rpc::optional<std::optional<replica::exception_variant>>> result_hit_rate) {
+                auto&& [result, hit_rate, opt_exception] = result_hit_rate;
+                if (opt_exception.has_value() && opt_exception->has_value()) {
```
Somewhat sad but can only be fixed in rpc. @gleb-cloudius we should have a way to pass default-constructible optional values, so that if they're missing they are replaced by a default-constructed value rather than an optional.

gleb-cloudius replied:
You can do it in out IDL for new members in structs since there is a syntax for it (you can even specify default value). For types that are directly managed by rpc it is not clear how to specify what you want. May be new rpc type rpc::optional_with_default_value(). Not sure it worth it to remove the `if`.

I am also not sure why do we need `std::optional` inside `rpc::optional` here. Doesn't `replica::exception_variant` have "empty" value?

avikivity replied:
We could just say that if the message ends before an argument is deserialized, it is default-constructed (if possible). Just like we say it for structs.

gleb-cloudius replied:
And if it is not? Fail compilation? Or do different things in compile time based on if `T` is default constructable? But then rpc may change without notice by changing T without doing any IDL changes. Also some code may want to know if a value was provided even if `T` is default construcatble and there should be a way to express it.

IIRC for structs we require to specify default value in IDL to not get a value as rpc::optional.

gleb-cloudius replied:
Also RPC has to know what arguments are optional because it needs to be able to detect that data is missing. Suppose we have an RPC call `foo(bool, bool)`. If a fragment has only enough data for one `bool` we should signal underflow. OTOH if we had `bar(bool)` and later extended it to have `bar(bool, bool)` having only one `bool` in a networking frame is not an error.

avikivity replied:
How about this:
- new arguments have to be wrapped in rpc::optional in IDL
- we provide three options for the C++ handler signature:
   1. std::optional (disengaged if argument not provided)
   2. initializer in IDL (used to initialize if argument not provided)
   3. default constructor (requires that T is default constructible)
  
IIUC these changes would all be in IDL, not RPC.

avikivity commented on db/per_partition_rate_limit_extension.hh:
```diff
@@ -0,0 +1,42 @@
+        return ser::serialize_to_buffer<bytes>(_options.to_map());
+    }
+    static std::map<sstring, sstring> deserialize(const bytes_view& buffer) {
+        return ser::deserialize_from_buffer(buffer, boost::type<std::map<sstring, sstring>>());
```
Does this mean that this shows up in the schema as a blob?

I prefer a traditional serialization into a dedicated map column, so the schema can still be understood by people.

piodul replied:
Yes, it's serialized as a blob:

```
cqlsh> select extensions from system_schema.tables where table_name = 't' allow filtering;

 extensions
--------------------------------------------------------------------------------------------------------
 {'per_partition_rate_limit': 0x01000000140000006d61785f72656164735f7065725f7365636f6e6403000000313030}

(1 rows)
```

The type of the `extensions` column is `frozen<map<text, blob>>`. Even if we serialize it e.g. as a json, cqlsh will show ascii bytes. Should I change the serialization, or rather keep it as it is for consistency (all existing schema extensions use `ser::deserialize_from_buffer`)?

avikivity replied:
The question is whether this should be a schema extension or not. Most times when we extend the schema we add columns.

psarna replied:
Schema extensions bypass the need of introducing new schema features at a (usually negligible) runtime cost, so I'm generally a fan, as long as parsing data from these features is not on the hot path.

psarna replied:
By the way, if we take https://github.com/scylladb/scylla/issues/9571 into consideration, it will become the server's responsibility to produce DESCRIBE results. Thanks to that, we'll be able to present all such blobs in human-readable form.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -4864,7 +4866,7 @@ future<bool> storage_proxy::cas(schema_ptr schema, shared_ptr<cas_request> reque
 
                 auto pr = partition_ranges; // cannot move original because it can be reused during retry
-                auto cqr = co_await query(schema, cmd, std::move(pr), cl, query_options);
+                auto cqr = co_await query(schema, cmd, std::move(pr), cl, db::allow_per_partition_rate_limit::no, query_options);
```
Maybe we should move it to read_cmd? That's already an extension friendly structure and can eliminate lots of boilerplate (and runtime cost of passing parameters).

piodul replied:
I put `allow_limit` into `read_command`... However, it's a coordinator-only parameter (it is converted to `db::per_partition_rate_limit::info` during coordinator logic), so it it resulted in only slight reduction of boilerplate (it's not serialized in rpc, by the way).

avikivity commented on replica/database.cc:
```diff
@@ -1335,6 +1335,33 @@ request_class classify_request(const database_config& _dbcfg) {
+        .replicas_in_group = 1,
+        .needed_for_cl_in_group = 1,
+    };
+    column_family& cf = find_column_family(s);
```
Should it not be a method of `table`? Then a caller who already knows the table can avoid a lookup.

piodul replied:
It needs `_rate_limiter` and `_dbcfg` from `database`, I would have to expose them to the outside callers. Instead I changed this method to take `table&` as an argument.

I hoped to eliminate one call to `find_column_family` this way, but later I deduplicated the code responsible for calculating the right rate limiter info and, unfortunately, the lookup is always there.

avikivity commented on docs/design-notes/per-partition-rate-limit.md:
```diff
@@ -0,0 +1,229 @@
+# Per-partition rate limiting
+
+Scylla clusters operate best when the data is spread across a large number
```
Please also add user-facing documentation in cql-extensions.md.

piodul replied:
Done

avikivity commented:
Looks good. Please add read/write perf_simple_query results to the cover letter.

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -4995,14 +5006,17 @@ storage_proxy::handle_write(netw::messaging_service::msg_addr src_addr, rpc::opt
                         // ignore timeouts so that logs are not flooded.
                         // database total_writes_timedout counter was incremented.
                         l = seastar::log_level::debug;
+                    } else if (std::holds_alternative<replica::rate_limit_exception>(errors.local.reason)) {
```
Why not adding `||` to the `if`?

piodul replied:
OK, done.

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -5218,8 +5269,21 @@ storage_proxy::handle_read_digest(const rpc::client_info& cinfo, rpc::opt_time_p
+                        }
+                    }
+                    return make_exception_future<final_tuple_type>(std::move(eptr));
+                }
```
We have almost exactly same code three times. I wonder if we can put it in a function. Types are a little bit different, so it may be tricky and not worth it.

piodul replied:
Done. I think it didn't turn out too bad.

gleb-cloudius commented on idl/per_partition_rate_limit_info.idl.hh:
```diff
@@ -0,0 +1,25 @@
+    uint16_t participating_groups;
+    uint16_t replicas_in_group;
+	uint16_t needed_for_cl_in_group;
+};
```
Is there some kind of design document that explains what the series is doing? Why do you need to pass this data to all replicas if I see a coordinator code that decides to  rate-limit? Do you decide on replicas as well? Why?

piodul replied:
> Is there some kind of design document that explains what the series is doing? Why do you need to pass this data to all replicas if I see a coordinator code that decides to rate-limit? Do you decide on replicas as well? Why?

There is a document in the design notes directory, added in the last commit (probably should have placed it in the first commit instead, my bad). It should answer your questions - you can start from the `Implementation` section and read to the end.

piodul replied:
Moved the commit introducing the document earlier in the series.

gleb-cloudius commented on docs/design-notes/per-partition-rate-limit.md:
```diff
@@ -0,0 +1,229 @@
+given that `RF` is usually not large we consider this behavior to be
+an acceptable limitation.
+
+### Optimization for a coordinator inside the replica set
```
Why is it an optimization and not the primary (and only) mode of operation? We do require users to use shard aware drivers and consider anything else as a bug.

piodul replied:
I wouldn't say that not using a shard-aware driver is a bug. Of course, using a shard-aware driver is preferable as shard-aware operations are cheaper, but we handle non-shard-aware operations correctly, too. As this is an overload protection feature I think it is important to cover all cases.

I chose "optimization" here because rejection is less costly in this case (no replica work is wasted).

gleb-cloudius replied:
May be not a bug, but definitely "do not expect a good performance" limitation. This and "not prepared statement" are the first thing that we check if there are performance problems.

avikivity replied:
We can say that rate limiting works best with a shard aware driver.

gleb-cloudius replied:
I think the problem exists for shard aware drivers as well. Suppose a coordinator decided to not fail the request, but on of the replicas still did. We will have a request that reached CL but generated hint + immediate replay.

piodul replied:
@gleb-cloudius No, when the coordinator makes the decision then the replicas do not make the decision. The behavior of replicas is controlled with `db::per_partition_rate_limit::info`, which is a variant that can hold:

1. `std::monostate` - replica does not count the operation and cannot reject it,
2. `db::per_partition_rate_limit::only_account` - replica accounts the operation internally but cannot reject it. Because of this, the replica's counter can be more accurate.
3. `db::per_partition_rate_limit::account_and_enforce` - replica accounts the operation and decides whether to reject or not.

When the coordinator is a replica, it first decides whether to proceed and if it does it uses 2. for remote replicas and 1. for the local operation.

When the coordinator is not a replica, then the coordinator does not decide and sends 3. to replicas.

gleb-cloudius replied:
OK. The shard aware behaviour is good then. Still the most complexity in the patchset is around replica<->coordinator communication which is relevant only for not shard aware drivers.

gleb-cloudius replied:
I think we should consider it now. Because post merge RPC ABI cannot be simplified, only extended.

piodul replied:
I thought about accounting given operation multiple times on the coordinator, but I'm worried about misconfigured/bugged drivers which are shard-aware but do not round-robin between replicas correctly. Such bugs do happen, for example we recently reported a bug in gocql where only the first replica is chosen for a given token ([link](https://github.com/gocql/gocql/issues/1621)). Accounting RF-times on the coordinator can result in the limit being applied too aggressively.

The current design makes it possible to avoid assumptions about the client's load balancing policy.

piodul replied:
As far as coordinator RPC is concerned, the idea with coordinator-local cache ([here](https://github.com/scylladb/scylla/pull/9810#discussion_r879498072)) might also make the RPC a bit simpler: coordinator send a single flag whether to rate limit or not, and replicas return information about the current load of the partition. It would be incompatible with what is implemented in this PR, though.

avikivity replied:
> I thought about accounting given operation multiple times on the coordinator, but I'm worried about misconfigured/bugged drivers which are shard-aware but do not round-robin between replicas correctly. Such bugs do happen, for example we recently reported a bug in gocql where only the first replica is chosen for a given token ([link](https://github.com/gocql/gocql/issues/1621)). Accounting RF-times on the coordinator can result in the limit being applied too aggressively.
> 
> The current design makes it possible to avoid assumptions about the client's load balancing policy.

Such drivers will increase the overload for hot partitions since there will be fewer coordinators to perform the rejection. So we can demand fixed drivers for this to work well.

The fact that the bug is new and open weakens the statement, because it means there's a good chance people will encounter it, but perhaps by the time this gets to a release the fixed driver will be available.

piodul replied:
I found another offender: the python driver. By default it uses a token/shard-aware strategy which always chooses the first replica. There is an option "shuffle_replicas" in the constructor which fixes the problem when set to `True`, but it is `False` by default.

https://github.com/scylladb/python-driver/blob/94b64bb5571ed9c47d8cb7e8e19fcb6806cf1f2f/cassandra/policies.py#L327-L330

I verified this behavior in dtests - and rate limitng indeed throttles too heavily with a default policy and accounting RF-times on the coordinator. We will have to change the default in the driver or tell users to explicitly set shuffle_replicas=True.

psarna replied:
@fee-mendes fyi ^

kbr- replied:
Note that using the same node for coordination is good for LWT liveness.

avikivity replied:
Ok. So we shouldn't do CL/RF accounting on the coordinator, unless the client promises to split the load evenly (and in fact it won't when a node is down).

avikivity replied:
ok

piodul commented on docs/design-notes/per-partition-rate-limit.md:
```diff
@@ -0,0 +1,229 @@
+of requests to a partition exceed configured limit, they start being rejected
+with such a probability that, statistically, the number of accepted reject
+falls back to the configured limit.
+
```
> I see problems with rejecting on replicas. First of all it may causes a lot of work by a coordinator to be wasted. For writes we send to all replicas, so if some reject we may still complete the operation successfully, but for reads we contact only CL with probably one replica to retry with and the chance for operation to fail becomes large while a coordinator already wasted a lot of effort.

Yes, when the coordinator is not a replica replicas will do some wasted work. However, even still less work is actually done compared to not rate limiting at all because replicas sometimes will reject - so the total work the replica shards have to do will be lower than without rate limiting. I'm not sure how it could be done better - in this design, replicas are responsible for tracking partitions that they own, so letting replicas decide avoids the need for additional round trips.

Note that this problem does not happen for shard-aware drivers as in this case the coordinator usually _is_ a replica and it is allowed to decide by itself whether to reject or not. In this case, the coordinator either proceeds with the operation and replicas are not allowed to reject it, or the coordinator rejects the operation and does not contact the replicas.

> 
> Also what happens if some replicas rejects a write but it still succeeds? Do we write hints in this case? Do we reply hints with writes that can be rejected? Because if not we achieved nothing, just delayed the write a little.

A write is considered failed if less then `CL` replicas accept. We write hints only for writes which succeed. So, if `L` ops/s is the limit, then we will write at most `L * (RF - CL)` additional hints/s. Again, those hints will only be generated when _not_ using a shard-aware driver.

gleb-cloudius replied:
> > I see problems with rejecting on replicas. First of all it may causes a lot of work by a coordinator to be wasted. For writes we send to all replicas, so if some reject we may still complete the operation successfully, but for reads we contact only CL with probably one replica to retry with and the chance for operation to fail becomes large while a coordinator already wasted a lot of effort.
> 
> Yes, when the coordinator is not a replica replicas will do some wasted work. However, even still less work is actually done compared to not rate limiting at all because replicas sometimes will reject - so the total work the replica shards have to do will be lower than without rate limiting. I'm not sure how it could be done better - in this design, replicas are responsible for tracking partitions that they own, so letting replicas decide avoids the need for additional round trips.

As I said in some other comment I am not so sure non shard aware driver is a case worth covering especially as the solution is far from been perfect. @avikivity?

> 
> Note that this problem does not happen for shard-aware drivers as in this case the coordinator usually _is_ a replica and it is allowed to decide by itself whether to reject or not. In this case, the coordinator either proceeds with the operation and replicas are not allowed to reject it, or the coordinator rejects the operation and does not contact the replicas.
> 
> > Also what happens if some replicas rejects a write but it still succeeds? Do we write hints in this case? Do we reply hints with writes that can be rejected? Because if not we achieved nothing, just delayed the write a little.
> 
> A write is considered failed if less then `CL` replicas accept. We write hints only for writes which succeed. So, if `L` ops/s is the limit, then we will write at most `L * (RF - CL)` additional hints/s. Again, those hints will only be generated when _not_ using a shard-aware driver.

So in that case you actually multiplied the work you are doing. You add load to the coordinator and do not drop any load from replicas since hints will be replayed shortly. I cannot say how rare on otherwise this case is (given that probabilities are involved in each replica's decision).

piodul replied:
> > > Also what happens if some replicas rejects a write but it still succeeds? Do we write hints in this case? Do we reply hints with writes that can be rejected? Because if not we achieved nothing, just delayed the write a little.
> > 
> > 
> > A write is considered failed if less then `CL` replicas accept. We write hints only for writes which succeed. So, if `L` ops/s is the limit, then we will write at most `L * (RF - CL)` additional hints/s. Again, those hints will only be generated when _not_ using a shard-aware driver.
> 
> So in that case you actually multiplied the work you are doing. You add load to the coordinator and do not drop any load from replicas since hints will be replayed shortly. I cannot say how rare on otherwise this case is (given that probabilities are involved in each replica's decision).

Yes, I do more work for accepted writes, but I also do less work for rejected writes. We do not write hints for failed writes. For example, if the limit is 100 ops/s for a partition and a client starts sending 10k ops/s, with CL=QUORUM and RF=3 we will generate 100 hints for 100 successful writes but we won't generate any for the remaining 9900. The overall load will be lower.

One thing I should probably mention which might be not obvious here: if a throughput for a partition is kept below the limit, then replicas will never reject. No hints will be generated. We do not pay the cost of hints if the user behaves nicely.

There is probably a range between `limit` and `limit * k` where `k` is a small constant larger than one, such that the load is higher than if rate limiting was disabled. I'm not sure if it is a big problem - the limit should be a small fraction of what the shard can handle (otherwise there won't be much resources left for other shards when rate limiting kicks in). I would expect that in the case of a "hot partition" problem, the number of overloaded partitions is small.

gleb-cloudius replied:
> > > > Also what happens if some replicas rejects a write but it still succeeds? Do we write hints in this case? Do we reply hints with writes that can be rejected? Because if not we achieved nothing, just delayed the write a little.
> > > 
> > > 
> > > A write is considered failed if less then `CL` replicas accept. We write hints only for writes which succeed. So, if `L` ops/s is the limit, then we will write at most `L * (RF - CL)` additional hints/s. Again, those hints will only be generated when _not_ using a shard-aware driver.
> > 
> > 
> > So in that case you actually multiplied the work you are doing. You add load to the coordinator and do not drop any load from replicas since hints will be replayed shortly. I cannot say how rare on otherwise this case is (given that probabilities are involved in each replica's decision).
> 
> Yes, I do more work for accepted writes, but I also do less work for rejected writes. We do not write hints for failed writes. For example, if the limit is 100 ops/s for a partition and a client starts sending 10k ops/s, with CL=QUORUM and RF=3 we will generate 100 hints for 100 successful writes but we won't generate any for the remaining 9900. The overall load will be lower.

I am not discussing a failed writes here. I am talking about writes that were suppose to fail because one of the replicas is overloaded (if the write would gi through that replica it would have failed) and asks to rate limit access, but because we "distributed" decision about the failure the write may actually succeed and overload the overloaded replica and the coordinator. Note that the issue is not only hint that is written but also write that happens when hint is replayed. So the replica tries to avoid the write, but didn't.  


> 
> One thing I should probably mention which might be not obvious here: if a throughput for a partition is kept below the limit, then replicas will never reject. No hints will be generated. We do not pay the cost of hints if the user behaves nicely.

This is obvious.

piodul replied:
> > > I see problems with rejecting on replicas. First of all it may causes a lot of work by a coordinator to be wasted. For writes we send to all replicas, so if some reject we may still complete the operation successfully, but for reads we contact only CL with probably one replica to retry with and the chance for operation to fail becomes large while a coordinator already wasted a lot of effort.
> > 
> > 
> > Yes, when the coordinator is not a replica replicas will do some wasted work. However, even still less work is actually done compared to not rate limiting at all because replicas sometimes will reject - so the total work the replica shards have to do will be lower than without rate limiting. I'm not sure how it could be done better - in this design, replicas are responsible for tracking partitions that they own, so letting replicas decide avoids the need for additional round trips.
> 
> As I said in some other comment I am not so sure non shard aware driver is a case worth covering especially as the solution is far from been perfect. @avikivity?

I think that we do care about users which use non-shard-aware drivers.

Now that I think about it, an idea comes to my mind how to avoid waste: replicas could inform coordinators about the level of overload given partition has (a bit like the view backlog). This information would be sent only when the partition is overloaded and coordinators would keep it in a small cache which expires over time. Then, a coordinator could use information in its local cache to probabilistically reject. The main problems seem to be choosing the cache size and maybe that the information in the cache can be a bit stale which could affect the accept rates.

Of course, this would require changing this PR quite a bit and seeing if this idea works alright (I would estimate 2-4 days of work), so this should also be considered while deciding whether to do it now, later, or at all.

WDYT @avikivity @gleb-cloudius ? Any other ideas?

avikivity replied:
For the majority of queries, won't all replicas reach the same decision? For example, if the rate limit is 100/s and people are sending us 1000 queries/s. then for some queries we may get mismatched responses but for at least 700 all replicas will fail?

piodul replied:
The probability is not entirely trivial to compuate, so I ran an experiment: RF=3, CL=ALL, limit 100/s, workload 1000 reads/s. Each replica reported a bit above 400 successful reads/s.

- 59% cases had 0 successes
- 21% cases had 1 success
- 6% cases had 2 successes
- 14% cases had 3 successes (higher than 10% because of the coordinator optimization)

So, in about 59% + 14% = 73% cases all participating replicas made the same decision.

avikivity replied:
I am super pleased that my 70% guess was so close to reality.

In the future we can try to coordinate the failures. If the algorithm is probabilistic, we can generate a random number on the coordinator and send it to the replicas, so they'll make similar decisions if their states are similar.

gleb-cloudius replied:
> The probability is not entirely trivial to compuate, so I ran an experiment: RF=3, CL=ALL, limit 100/s, workload 1000 reads/s. Each replica reported a bit above 400 successful reads/s.
> 
>     * 59% cases had 0 successes
> 
>     * 21% cases had 1 success
> 
>     * 6% cases had 2 successes
> 
>     * 14% cases had 3 successes (higher than 10% because of the coordinator optimization)
> 
> 
> So, in about 59% + 14% = 73% cases all participating replicas made the same decision.

In real life with more networking delays and node restarts (loosing statistics) we may have much more cases where replicas disagree. Even in this simple test there are 6% requests that will increase local on a system instead of decreasing it.

avikivity replied:
Why do the 6% cases increase load? They decreased it, though not as much as they could have.

avikivity replied:
btw, I think that the coordinator decision should account for RF (write) or CL (read) attempts, not 1, since that's the amount of work it saved. That balances out whether we had a coordinator or replica decision.

avikivity replied:
Ah, that's why account_only and account_and_enforce. But maybe RF-accounting at the coordinator is simpler.

gleb-cloudius replied:
> Why do the 6% cases increase load? They decreased it, though not as much as they could have.

In those 6% of the cases a write query will succeed and thus for one replica that replied with an error a hint will be stored. Said hint will be replied shortly (since the node is alive), but this time with forced apply. This result in more work, not less for non trivial amount of requests even in a simple test.

avikivity replied:
Yes. Well it's unfortunate. We should try to get replicas to arrive at the same decision (by sending them the random number for a decision), but we can do it post-merge.

piodul replied:
I implemented the idea with the coordinator sending a random number to replicas. As a result, probability calculation is now much simpler.

avikivity replied:
> I'll test cluster upgrade tomorrow. I also measured a regression of 600 insns/op for writes and 500 for reads, I'll investigate where it comes from and see what can be done about it.

First, check if there is a regression in the number of tasks/allocations.

piodul commented:
I ran the write benchmark on some of the commits and measured an increase in instruction count in the following ones:

- a14932f48840b6736835e0fa00bf22676b94883b (`storage_proxy: resultize return types of write handler creation path`): 80 insns
  This commit changes a bunch of functions to return `result<>`, so there may be some overhead related to checking whether the function succeeded or not.
- 5051a184d229abac30b1c0fbb5fd1f64d62ea7bc (`storage_proxy: choose the right per partition rate limit info in write handler`): 90 insns
  This commit adds a call to `choose_rate_limit_info` in `create_write_response_handler_helper`. I'll try adding a check which detects that the table is not rate limited and returns early from that function
- 5c9612219af98731fdf3a6f9cb9b9f974998f65c (`storage_proxy: add allow rate limit flag to mutate/mutate_result`): _350_ insns (!)
  I'm a bit confused about this one. It looks innocent as it only adds a parameter to a bunch of methods and to an execution stage, yet it introduces the biggest regression. There are other commits in this PR which add a similar parameter and they are not as costly, so maybe the execution stage is at fault?

piodul commented:
> First, check if there is a regression in the number of tasks/allocations.

@avikivity There was no regression in tasks/allocations (see the cover letter).

psarna commented:
> I ran the write benchmark on some of the commits and measured an increase in instruction count in the following ones:
> 
> * [a14932f](https://github.com/scylladb/scylla/commit/a14932f48840b6736835e0fa00bf22676b94883b) (`storage_proxy: resultize return types of write handler creation path`): 80 insns
>   This commit changes a bunch of functions to return `result<>`, so there may be some overhead related to checking whether the function succeeded or not.
> * [5051a18](https://github.com/scylladb/scylla/commit/5051a184d229abac30b1c0fbb5fd1f64d62ea7bc) (`storage_proxy: choose the right per partition rate limit info in write handler`): 90 insns
>   This commit adds a call to `choose_rate_limit_info` in `create_write_response_handler_helper`. I'll try adding a check which detects that the table is not rate limited and returns early from that function
> * [5c96122](https://github.com/scylladb/scylla/commit/5c9612219af98731fdf3a6f9cb9b9f974998f65c) (`storage_proxy: add allow rate limit flag to mutate/mutate_result`): _350_ insns (!)
>   I'm a bit confused about this one. It looks innocent as it only adds a parameter to a bunch of methods and to an execution stage, yet it introduces the biggest regression. There are other commits in this PR which add a similar parameter and they are not as costly, so maybe the execution stage is at fault?

The first few (6?) parameters would customarily be passed via registers, but the rest goes on the stack. Maybe that contributed? Although 350 surplus instructions sounds a little vast.

psarna commented:
I double-checked that there's indeed a similar regression caused by [5c96122](https://github.com/scylladb/scylla/commit/5c9612219af98731fdf3a6f9cb9b9f974998f65c) on my machine too

michoecho commented:
> /cc @michoecho

If you want to know what exactly causes the difference, then PT is the tool for the job. You can collect a trace, use https://github.com/michoecho/perf2perfetto to pick the interesting timespan, disassemble the instructions executed within the timespan, and e.g. compare the outputs for "before" and "after" with a difftool.

piodul commented:
> If you want to know what exactly causes the difference, then PT is the tool for the job. You can collect a trace, use https://github.com/michoecho/perf2perfetto to pick the interesting timespan, disassemble the instructions executed within the timespan, and e.g. compare the outputs for "before" and "after" with a difftool.

Unfortunately, my developer machine has an AMD CPU, and AFAIK PT is only supported on Intel...

avikivity commented:
> I tried grouping some of the parameters (four of them) on the write path into one struct, but I only managed to save 50 cycles this way. Link to the branch: [piodul@8def543](https://github.com/piodul/scylla/commit/8def5438cda2e879d1233ff71a69bcb639b42397)
> 
> I'll have to come up with something else, or try profiling it in some other way than by using PT...

Should pass by reference, otherwise there's still a large inlining load ( + copying if inlining fails).

It's tricky to pass by reference since you need to guarantee liveness. In the past we've discussed making service::permit (coordinator) or reader_permit (replica) a bag of stable things needed for a query) (and then rename it into something other than a permit).

piodul commented:
This change helped: https://github.com/piodul/scylla/commit/b37b6084d8920b6ad8ca66216dc61837274114ee. I inlined a bit of the `choose_rate_limit_info` function and added a check which quickly detects if rate limiting is disabled for the table and skips the function completely.

I think the cause was more prozaic than we suspected: before 5c9612219af98731fdf3a6f9cb9b9f974998f65c `allow_limit` was always `no` and the function returned quickly. After that commit, it was usually `yes` and `choose_rate_limit_info` just becomes more expensive which explains the regression. My fix makes the function cheaper in case the table has rate limiting disabled.

After the fix, the regression is 200 ops for writes and 150 ops for reads. I'll apply the fix to the relevant commits.

avikivity commented:
That's acceptable, it's 0.5% for something that protects throughput (exceptionless timeouts) and an important new feature.

btw, say something about exceptionless timeouts in the cover letter and provide a demonstration (benchmark), for many workloads it's even more important than the new feature.

gleb-cloudius commented on docs/design-notes/per-partition-rate-limit.md:
```diff
@@ -0,0 +1,142 @@
+
+The assumption is that all replicas will converge to similar counter values.
+Most of the time they will agree on the decision and not much work
+will be wasted due to some replicas accepting and other rejecting.
```
Can you rerun the test you ran with previous series to see how many replicas reject the write ?

piodul replied:
I added a feature which synchronizes the rate limiter tick with the system time. The results of the test will now be highly variable on how well the nodes' clocks are synchronized.

I re-ran the test described [here](https://github.com/scylladb/scylla/pull/9810#discussion_r879643131). All nodes were running on the same machine, and additionally I disabled the "coordinator optimization" so that it does not influence the results. In 99.8% cases replicas were in agreement.

We can estimate the worst case. Basing the calculations on [this](https://github.com/scylladb/scylla/pull/9810#issuecomment-1155071509) and assuming a steady request rate of `V`, replicas will be in agreement if the random number sent by coordinator is _outside_ the range `[P(2V), P(V)]`. The chance of _disagreement_ is at most:

```
P(V) - P(2V)
= (L/V) * (1 / ln 2 - 1 / (2 * ln 2))
= (L/V) * (1 / (2 * ln 2))
~= (L/V) * 0.7213475204444817
```

So, `0.72 * L` out of `V` requests per second may have disagreeing replicas.

gleb-cloudius replied:
What is 'L' her? Also will reboot of a node affect the result since it loses it local view of the rate limit?

piodul replied:
Yes, temporarily. The per-partition counters are not preserved and are reset to 0 on reboot. Until the new node "heats up" and its counter value stabilizes between V and 2V, it will serve more requests. The counter converges exponentially, from what I noticed it takes about 5 seconds - then, assuming that the clocks are still synchronized after reboot, the agreement rate should be the same.

gleb-cloudius replied:
So suppose we have a scenario when all nodes were rate limiting and two were rebooted. Now those two start to accept but one still rejects and thus requests succeed, but we are in a situation we try to avoid when hints are stored and immediately replayed.

piodul replied:
Yes, the newly restarted nodes will accumulate some hints for a short while until they start agreeing, then they will be replayed to the old node. I'm not sure how we can avoid it when replicas independently decide whether to reject or not.

/cc @avikivity @psarna

piodul replied:
We could try solving it by gathering information about the counters from replicas, caching them and decaying on the coordinator, like mentioned [here](https://github.com/scylladb/scylla/pull/9810#discussion_r879498072). The question is whether we want to do it and if we want to do it now. If it is implemented, then some of the current parts of the solution will become obsolete (e.g. typed exceptions over RPC).

avikivity replied:
I think it's okay if we generate some noise on restarts.

gleb-cloudius replied:
Back when we wrote HWLB we were very careful to avoid noise on restart. I wouldn't want to complicate the solution to remove the noise either though. IMO, which I already stated before, the cleanest solution would be to support topology aware drivers only since we pretty much require using them for best performance. This would have simplified the patch series greatly as well.

piodul commented:
> btw, say something about exceptionless timeouts in the cover letter and provide a demonstration (benchmark), for many workloads it's even more important than the new feature.

@avikivity Timeouts are not exceptionless yet. Some replica exceptions are still thrown and rethrown on the coordinator. Changes made to the RPC only make sure that the type of the rate limit exception is preserved across RPC.

For shard-aware drivers the current implementation doesn't throw exceptions, as the rejects are only done on the coordinator and it uses `bo::failure(rate_limit_exception())` for that.

After this PR, I'd like to continue work on https://github.com/scylladb/scylla/pull/10368. With it, getting rid of remaining throws should be easy. If this approach is not accepted, then I'll resultize the remaining parts and we will have to allow seastar rpc not to return exceptions (changing signatures in our RPC definitions won't help as timeouts are generated by the seastar rpc client).

avikivity commented on schema.cc:
```diff
@@ -1302,6 +1309,11 @@ schema_builder& schema_builder::with_tombstone_gc_options(const tombstone_gc_opt
 }
 
+schema_builder& schema_builder::with_per_partition_rate_limit_options(const db::per_partition_rate_limit_options& opts) {
+    add_extension(db::per_partition_rate_limit_extension::NAME, ::make_shared<db::per_partition_rate_limit_extension>(opts));
```
How does it work in mixed clusters? Suppose we call this on an upgraded node, what will an old node think of the new extension?

Perhaps we need to restrict this until the cluster is fully updated.

piodul commented:
While trying to calculate the replica waste as requested by Gleb here: https://github.com/scylladb/scylla/pull/9810#discussion_r895145391, I realized that the pattern with the coordinator choosing a random number introduces more inaccuracies which are not that insignificant.

Let `L` - limit ops/s, `V` - current rate of ops/s, `C` - current counter value which oscillates between `V` and `2V`. Currently, replicas calculate a threshold `P(C) = L / C * ln 2` and accept if the random number sent by coordinator is below the threshold. This works well in single replica case, but not necessarily with multiple replicas. With CL=ONE, the actual threshold for the whole operation is the __maximum__ of all replica thresholds, and for CL=ALL it is the __minimum__. The rate limiter timers are not synchronized across replicas, so the time when the counter is halved might be different on all replicas and taking the minimum/maximum might have significant effect.

I ran a test similar to https://github.com/scylladb/scylla/pull/9810#discussion_r879643131, but additionally I disabled the "coordinator optimization" so that the replicas always decide whether to reject. I noticed that the cluster only accepted about 85 requests per second instead of the configured 100.

We can estimate how big the error can be. There are two extremes: `P(V)` and `P(2V)`. We can calculate the deviation from the ideal `L/V` ratio of accepted requests:

- `L/V - P(2V) = (L/V) * (1 - 1 / (2 * ln 2)) ~= (L/V) * 0.2786524795555183` => 28% more rejected
- `P(V) - L/V = (L/V) * (1 / ln 2 - 1) ~= (L/V) * 0.4426950408889634` => 44% more accepted

So, the extremes are either 28% more requests rejected or 44% more requests accepted than necessary.

piodul commented:
v2 -> v3:
- Applied the performance fixes to the right commits, updated the `perf_simple_query` benchmark results in the cover letter.
- Creation of tables with per-partition rate limiting is disallowed until all nodes are upgraded.
- Fixed the logic which decides whether operation should be rate limited or not (I broke it while doing performance optimizations) and added a test for it.
- Re-added `account_only` (for situations where the coordinator decides whether to reject and then only tells nodes to account).
- The rate limiter now synchronizes its ticks with the system time. The rationale for this is that the scheme where the coordinator chooses a random number and replicas calculate the threshold has better accurracy when the replicas have synchronized ticks (see the explanation [here](https://github.com/scylladb/scylla/pull/9810#issuecomment-1155071509)).
- Updated documentation about the rate limiting so that it mentions the inaccurracies that are caused by the unsynchronized clocks.
- Changed the wording slightly in docs around the new CQL error (RateLimitExceeded -> "rate limit exceeded").
- Removed one trace message (introduced previously in this PR) which is added when rate limiting is not enabled.
- Squashed a commit which introduced a bunch of functions in storage_proxy ([storage_proxy: add helper functions for calculating rate limit parame…](https://github.com/scylladb/scylla/pull/9810/commits/d9233fc990dfa662ce479f0bb954faa673a27095)) with a commit which uses them for the first time in order to avoid compilation warnings.

avikivity commented on gms/feature_service.hh:
```diff
@@ -108,6 +108,7 @@ public:
     gms::feature tombstone_gc_options { *this, "TOMBSTONE_GC_OPTIONS"sv };
     gms::feature parallelized_aggregation { *this, "PARALLELIZED_AGGREGATION"sv };
     gms::feature keyspace_storage_options { *this, "KEYSPACE_STORAGE_OPTIONS"sv };
+    gms::feature typed_errors_in_read_rpc { *this, "TYPED_ERRS_IN_READ_RPC"sv };
```
s/ERRS/ERRORS/ please. There's no reason to abbreviate, and it's inviting confusion when someone does a `git grep -i typed_errs_in_read_rpc` without realizing it's different.

piodul replied:
I abbreviated it because dtests look for the "ERROR" string in the logs on teardown. The feature name is printed to the logs when the node starts and most tests failed. Here is a CI run which failed due to this reason: https://jenkins.scylladb.com/job/releng/job/Scylla-CI/92/

I'll send a PR to scylla-ccm to improve the error detection in the logs. Maybe we can also change the name of the feature and align the variable name with the string - e.g. `TYPED_EXCEPTIONS_IN_READ_RPC`?

avikivity replied:
I'm okay with either name.

avikivity commented:
Looks good. I left the most superficial comment ever.

Please update the cover letter with information about how you tested mixed clusters. I want reads and writes to hit both old and new coordinators and replica calls to hit both old and new replicas.

@gleb-cloudius please review at least to storage_proxy part.

piodul commented:
v2 -> v2.1:
- rebased
- changed the feature name to `TYPED_ERRORS_IN_READ_RPC`

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/990/

piodul commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/990/

CI failed due to some timeouts in the unit tests. The CI run for the previous version (v2) also had a timeout, although only one. I don't see similar timeouts in other CI runs (apart from this one, maybe: https://jenkins.scylladb.com/job/releng/job/Scylla-CI/980/), but at the same time I'm not sure how they are related to the PR. I'm trying to reproduce locally but with no luck so far.

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -2795,6 +2798,9 @@ void storage_proxy::send_to_live_endpoints(storage_proxy::response_id_type respo
             try {
                 std::rethrow_exception(eptr);
+            } catch (replica::rate_limit_exception&) {
+                // There might be a lot of those, so ignore
```
Should we count them?

piodul replied:
Yeah, if we are counting timeouts then probably it's a good idea to count rate limit exceptions, too. I'll add metrics, one for reads and one for writes.

piodul replied:
Added replica- and coordinator-side metrics. The coordinator actually has two metrics, separated by a label, one counts cases where the coordinator rejected and the other counts the cases where replicas made the decision to reject.

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -4995,14 +5006,15 @@ storage_proxy::handle_write(netw::messaging_service::msg_addr src_addr, rpc::opt
-                    if (is_timeout_exception(eptr)) {
-                        // ignore timeouts so that logs are not flooded.
+                    if (is_timeout_exception(eptr) || std::holds_alternative<replica::rate_limit_exception>(errors.local.reason)) {
+                        // ignore timeouts and rate limit exceptions so that logs are not flooded.
```
Here as well, the comment below claims that for timeouts a counter was incremented. What about rate limit?

piodul replied:
There is now an appropriate counter for rate limit exceptions and it is incremented. Updated the comment so that it mentions it.

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -3313,8 +3329,16 @@ class data_read_resolver : public abstract_read_resolver {
+    void on_error(gms::inet_address ep, error_kind kind) override {
+        switch (kind) {
+        case error_kind::RATE_LIMIT:
+            fail_request(exceptions::rate_limit_exception(_schema->ks_name(), _schema->cf_name()));
```
Shouldn't rate limit exception indicate if it was write or read?

piodul replied:
I'll add an enum indicating the operation type and will include that information in the error description. I also considered different exception types for read rate limit and write rate limit, but I think it will be cumbersome to add more rate limit types using this convention if we wish to do so in the future (e.g. for LWT).

piodul replied:
Done, added a field which describes the type of the operation. Additionally, I also included information whether an operation was rate limited on the coordinator or not (coordinator rejects are currently guaranteed not to modify the data so this case might be interesting to the users).

piodul commented:
v3 -> v4:
- Extracted `db::per_partition_rate_limit_options::operation_kind` into a separate enum: `db::operation_type`.
- The rate limit exception now contains information about the operation type (read or write) and information on whether the coordinator made the decision to reject or not.
- Added `no_exception` to `exception_variant` and got rid of `std::optional` around `replica::exception_variant`.
- Added metrics:
	- `scylla_storage_proxy_coordinator_write_rate_limited` - counts rejected writes due to per-partition rate limiting. Cases when coordinator decided to reject and when replicas decided to reject are counted separately (different labels).
	- `scylla_storage_proxy_coordinator_read_rate_limited` - like the previous one, but for reads.
	- `scylla_database_total_writes_rate_limited` - counts writes rejected by this replica.
	- `scylla_database_total_reads_rate_limited` - like the previous one, but for reads.
- Ran the manual upgrade test again and it seems to be working fine.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1001/
